crates/lair_keystore_api/src/sodium_secretstream.rs

```diff
@@ -0,0 +1,301 @@
+        AsS3Sender::get_dec_ctx_key(&*self.0)
+    }
+
+    /// Shutdown the channel.
+    pub fn shutdown(
+        &self,
+    ) -> impl Future<Output = LairResult<()>> + 'static + Send {
+        AsS3Sender::shutdown(&*self.0)
+    }
+}
+
+/// SodiumSecretStream - Receiver
+pub struct S3Receiver<T>(pub Box<dyn AsS3Receiver<T>>);
+
+impl<T> Stream for S3Receiver<T>
+where
+    T: for<'de> serde::Deserialize<'de>,
+{
+    type Item = LairResult<T>;
+
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        Stream::poll_next(std::pin::Pin::new(&mut self.0), cx)
+    }
+}
+
+/// Create a new S3 send/receive pair.
+pub fn new_s3_pair<T, S, R>(
+    send: S,
+    recv: R,
+    is_srv: bool,
+) -> impl Future<Output = LairResult<(S3Sender<T>, S3Receiver<T>)>> + 'static + Send
+where
+    T: 'static + serde::Serialize + for<'de> serde::Deserialize<'de> + Send,
+    S: 'static + tokio::io::AsyncWrite + Send + Unpin,
+    R: 'static + tokio::io::AsyncRead + Send + Unpin,
+{
+    async move {
+        // box these up into trait objects so we can easily refer to their types.
+        let mut send: PrivRawSend = Box::new(send);
+        let mut recv: PrivRawRecv = Box::new(recv);
+
+        // perform a key exchange with the remote.
+        let (tx, rx) = priv_kx(&mut send, &mut recv, is_srv).await?;
+
+        // perform a secretstream init handshake with the remote.
+        let (enc, dec) =
+            priv_init_ss(&mut send, tx.clone(), &mut recv, rx.clone()).await?;
+
+        // initialize framing so we know when we've got complete messages.
+        let (send, recv) = priv_framed(send, recv);
+
+        // wrap the streams with cryptography.
+        let (send, recv) = priv_crypt(send, enc, recv, dec);
+
+        // bundle up our output sender type.
+        let send: PrivSend<T> = PrivSend::new(send, tx, rx);
+        let send: S3Sender<T> = S3Sender(Arc::new(send));
+
+        // bundle up our output receiver type.
+        let recv: PrivRecv<T> = PrivRecv::new(recv);
+        let recv: S3Receiver<T> = S3Receiver(Box::new(recv));
+
+        Ok((send, recv))
+    }
+}
+
+// -- private -- //
+
+/// trait object type for AsyncWrite instance.
+type PrivRawSend = Box<dyn tokio::io::AsyncWrite + 'static + Send + Unpin>;
+
+/// trait object type for AsyncRead instance.
+type PrivRawRecv = Box<dyn tokio::io::AsyncRead + 'static + Send + Unpin>;
+
+mod framed;
+use framed::*;
+
+mod crypt;
+use crypt::*;
+
+mod inner;
+use inner::*;
+
+/// perform key exchange to generate secret rx key and secret tx key.
+fn priv_kx<'a>(
+    send: &'a mut PrivRawSend,
+    recv: &'a mut PrivRawRecv,
+    is_srv: bool,
+) -> impl Future<
+    Output = LairResult<(
+        sodoken::BufReadSized<{ sss::KEYBYTES }>,
+        sodoken::BufReadSized<{ sss::KEYBYTES }>,
+    )>,
+>
+       + 'a
+       + Send {
+    async move {
```

Comment:
Any reason to not just make this function async?

Reply:
yeah, there's no way to require the `Send` bound with just specifying `async fn` and the compiler errors end up in the wrong place, in the calling context without the bound.